### PR TITLE
sql : show backup info

### LINKED ISSUE #14219: backup/restore: ability to inspect a backup
given the path to a backup, it would be nice to be able to inspect it, to list the tables it contains and their schemas, time bounds, etc.

### PR DESCRIPTION
**Pull request by a6802739**
Fixes #14219. @dt

### CODE REVIEW

**Comment by dt** [comment]
This looks great!

A couple inline comments on specific bits of the code (add a parser test, switch to using endTime and include db names in results). Also, would be great to include an example of its output in the commit/PR message.

Also, a more general question of what our desired functionality is here though (mostly directed at  @danhhz):

as written, this lists all the tables in the backup and tells you what time the backup captured (with s/startTime/endTime below), one row per table -- together these seem like they're certainly the two most important things to know when looking at an opaque binary backup blob today.

That said, some attributes which might be interesting to also include are *per-backup*, like the timestamps, the data_size, the cluster and version that produced it, etc. I'm not quite sure how those fit into this list-of-tables result set -- maybe just duplicated on every row in the existing format? that could be kinda confusing if it makes them look like they're somehow per-table though?

Alternatively, maybe the results should be more free-form, e.g:
```
Field           |                 Value
---------------------------------------
time           |  2017-05-31...
data size    | 863mb
file names | blah.sst, foo.sst, bar.sst...
tables         | app.users, app.clients, app....
```

But that seems like a kinda clumsy way to represent the table lists though. maybe an entry per table with a repeated field name?

---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 473 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-KlUx4mitv7DpJPzyWxJ:-KlUx4mitv7DpJPzyWxK:bm3flsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/78d2454598f6f26eb3eb098952fcd87c1efb2e3b/pkg/ccl/sqlccl/backup.go#L473)):*
> ```Go
> 		return nil, nil, nil
> 	}
> 	/*
> ```

left over from debugging?

---

*[pkg/ccl/sqlccl/backup.go, line 592 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-KlUySX1xV4Oy7yH13Ow:-KlUySX1xV4Oy7yH13Ox:b-ick9ff) ([raw file](https://github.com/cockroachdb/cockroach/blob/78d2454598f6f26eb3eb098952fcd87c1efb2e3b/pkg/ccl/sqlccl/backup.go#L592)):*
> ```Go
> 		var ret []parser.Datums
> 		for _, descriptor := range desc.Descriptors {
> 			if table := descriptor.GetTable(); table != nil {
> ```

I think we want to include the db names too -- we'll want to iterate over the map once first to get the db descriptors make a map[ID]string, then below we can grab the dbs using `dbsByID[table.ParentID]`.

---

*[pkg/ccl/sqlccl/backup.go, line 595 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-KlV0CsRATMHEfxEZuVi:-KlV0CsRATMHEfxEZuVj:b-47vn2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/78d2454598f6f26eb3eb098952fcd87c1efb2e3b/pkg/ccl/sqlccl/backup.go#L595)):*
> ```Go
> 				temp := parser.Datums{
> 					parser.NewDString(table.Name),
> 					parser.MakeDTimestamp(time.Unix(0, desc.StartTime.WallTime), time.Second),
> ```

I think of the two timestamps, `endTime` is the more useful -- if we restored the backup, our DB would like it did as of the `endTime`, which is probably the more important of the two times to most casual inspections. 

We might also want to include startTime, which is interesting when trying to figure out how a bunch of interleaved backups fit together, but it isn't useful beyond that as far as I know... /cc @danhhz?

---

*[pkg/sql/parser/show.go, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-KlUxK2PeFxDcWjjqGgP:-KlUxK2PeFxDcWjjqGgQ:bw41h0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/78d2454598f6f26eb3eb098952fcd87c1efb2e3b/pkg/sql/parser/show.go#L49)):*
> ```Go
> // Format implements the NodeFormatter interface.
> func (node *ShowBackup) Format(buf *bytes.Buffer, f FmtFlags) {
> 	buf.WriteString("SHOW BACKUP ")
> ```

We generally like to include parser round-trip tests for every AST node, independent of the implementation testing we do elsewhere -- see the big of tests in `parse_test.go` -- the `BACKUP` related ones are around line 670ish.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-:-KlV0bvL-7gaKQoEINGe:b-uarevz)*

**Comment by danhhz** [comment]
Looks great overall! Thanks for the contribution!

I'm happy with the structure of this output. We could later add the ability to specify more than one backup to this command (say a full and some incrementals). It would validate whether they match and output the combined info. If the full backup had some tables the incremental one didn't, then the start and end times would need to be per table. (And to be clear, I'm not suggesting you do this in this PR.)

---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 592 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-KlUySX1xV4Oy7yH13Ow:-KlVF29OK-2RW6qDp0iP:b-i929iu) ([raw file](https://github.com/cockroachdb/cockroach/blob/78d2454598f6f26eb3eb098952fcd87c1efb2e3b/pkg/ccl/sqlccl/backup.go#L592)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think we want to include the db names too -- we'll want to iterate over the map once first to get the db descriptors make a map[ID]string, then below we can grab the dbs using `dbsByID[table.ParentID]`.
</blockquote></details>

+1 on this

---

*[pkg/ccl/sqlccl/backup.go, line 595 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-KlV0CsRATMHEfxEZuVi:-KlVE7qZbqpQaIPF530a:b-u8d6w3) ([raw file](https://github.com/cockroachdb/cockroach/blob/78d2454598f6f26eb3eb098952fcd87c1efb2e3b/pkg/ccl/sqlccl/backup.go#L595)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think of the two timestamps, `endTime` is the more useful -- if we restored the backup, our DB would like it did as of the `endTime`, which is probably the more important of the two times to most casual inspections. 

We might also want to include startTime, which is interesting when trying to figure out how a bunch of interleaved backups fit together, but it isn't useful beyond that as far as I know... /cc @danhhz?
</blockquote></details>

startTime will also be useful when we add the ability to restore to a time in the middle of a backup. I think we want both times

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-:-KlVEO5OUcDFTFq3XzFp:b-50qmbw)*

**Comment by a6802739** [comment]
I quite prefer each-table information per row. Here is the example:
```
root@:26660/test2> backup database test1, test2 to 'nodelocal:/Users/baidu/cockroachdb/src/github.com/cockroachdb/cockroach/backup';
+--------------------+-----------+--------------------+-------+
|       job_id       |  status   | fraction_completed | bytes |
+--------------------+-----------+--------------------+-------+
| 249821888626851840 | succeeded |                  1 |  3816 |
+--------------------+-----------+--------------------+-------+
(1 row)
root@:26660/test2> show backup 'nodelocal:/Users/baidu/cockroachdb/src/github.com/cockroachdb/cockroach/backup';
+----------+-------+---------------------------+---------------------------+
| database | table |        start_time         |         end_time          |
+----------+-------+---------------------------+---------------------------+
| test1    | temp1 | 1970-01-01 00:00:00+00:00 | 2017-06-01 09:39:50+00:00 |
| test1    | temp2 | 1970-01-01 00:00:00+00:00 | 2017-06-01 09:39:50+00:00 |
| test2    | temp3 | 1970-01-01 00:00:00+00:00 | 2017-06-01 09:39:50+00:00 |
+----------+-------+---------------------------+---------------------------+

```

---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 473 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-KlUx4mitv7DpJPzyWxJ:-KlXeqHbdqSSdaeb8PTP:b-64lgb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/78d2454598f6f26eb3eb098952fcd87c1efb2e3b/pkg/ccl/sqlccl/backup.go#L473)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

left over from debugging?
</blockquote></details>

Yes, my mistake.

---

*[pkg/ccl/sqlccl/backup.go, line 592 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-KlUySX1xV4Oy7yH13Ow:-KlXtGiZgO4SvhuaX5Wk:b-9755fa) ([raw file](https://github.com/cockroachdb/cockroach/blob/78d2454598f6f26eb3eb098952fcd87c1efb2e3b/pkg/ccl/sqlccl/backup.go#L592)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

+1 on this
</blockquote></details>

Thank you.

---

*[pkg/ccl/sqlccl/backup.go, line 595 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-KlV0CsRATMHEfxEZuVi:-KlXt4bXoppAZhYX7AeB:b-7vyt1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/78d2454598f6f26eb3eb098952fcd87c1efb2e3b/pkg/ccl/sqlccl/backup.go#L595)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

startTime will also be useful when we add the ability to restore to a time in the middle of a backup. I think we want both times
</blockquote></details>

I will add `end_time` then.

---

*[pkg/sql/parser/show.go, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-KlUxK2PeFxDcWjjqGgP:-KlXt9azD_EKCGQVRMl3:b-9755fa) ([raw file](https://github.com/cockroachdb/cockroach/blob/78d2454598f6f26eb3eb098952fcd87c1efb2e3b/pkg/sql/parser/show.go#L49)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

We generally like to include parser round-trip tests for every AST node, independent of the implementation testing we do elsewhere -- see the big of tests in `parse_test.go` -- the `BACKUP` related ones are around line 670ish.
</blockquote></details>

Thank you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-:-KlY0kM0vozYmhiMkoC8:b-hh80n3)*

File: pkg/ccl/sqlccl/backup.go [hunk 1]
```diff
@@ -20,6 +21,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/build"
 	"github.com/cockroachdb/cockroach/pkg/ccl/storageccl"
 	"github.com/cockroachdb/cockroach/pkg/ccl/utilccl"
+
```
Comment (dt) [inline pkg/ccl/sqlccl/backup.go]: Nit: looks like this snuck in here?

File: pkg/ccl/sqlccl/backup.go [hunk 2]
```diff
@@ -553,6 +557,65 @@ func backupPlanHook(
...
+	}
+	header := sqlbase.ResultColumns{
+		{Name: "database", Typ: parser.TypeString},
+		{Name: "table", Typ: parser.TypeString},
+		{Name: "start_time", Typ: parser.TypeTimestamp},
+		{Name: "end_time", Typ: parser.TypeTimestamp},
+	}
+	fn := func() ([]parser.Datums, error) {
+		str := toFn()
+		desc, err := readBackupDescriptor(baseCtx, str)
```
Comment (dt) [inline pkg/ccl/sqlccl/backup.go]: Not: I think we want to make our own span here, similar to above, and the pass that rather than baseCtx.

**Review by dt — APPROVED** [review]
Looks great!

A couple minor nits inline, and could you amend the commit message first line to match the PR title `sql : show backup info`?

**Comment by dt** [comment]
:lgtm: though I think it needs a rebase (and the ctx below)

---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 588 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-Kl8CoqdcQuYkqTpfKvz-r2-588:-KlZkRGCxzDPpklojM25:b-xy3gax) ([raw file](https://github.com/cockroachdb/cockroach/blob/37b0e239670e37c00b7f86cbec898a7568994840/pkg/ccl/sqlccl/backup.go#L588)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Not: I think we want to make our own span here, similar to above, and the pass that rather than baseCtx.
</blockquote></details>

almost: I think you want to pass `ctx` rather than `baseCtx` here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-:-KlZkB1rUetAdtcepaqF:bu6eclt)*

**Comment by a6802739** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/ccl/sqlccl/backup.go, line 588 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-Kl8CoqdcQuYkqTpfKvz-r2-588:-KlajcMStAEVjnlqMVGq:b6ac2wb) ([raw file](https://github.com/cockroachdb/cockroach/blob/37b0e239670e37c00b7f86cbec898a7568994840/pkg/ccl/sqlccl/backup.go#L588)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

almost: I think you want to pass `ctx` rather than `baseCtx` here.
</blockquote></details>

My mistake. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186)*

**Comment by dt** [comment]
:lgtm_strong: :thumbsup: 

---

Reviewed 4 of 5 files at r1, 1 of 2 files at r2, 1 of 1 files at r3.
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16186#-:-Kld-Ty2PrTg94Zr8bf_:biupebp)*

**Comment by a6802739** [comment]
@dt, So I could merge this now?

**Comment by dt** [comment]
Yep, looks good!

**Pull request merged**